File: openHAB/OpenHABNotificationsViewControllerTableViewController.m
//
//  OpenHABNotificationsViewControllerTableViewController.m
//  openHAB
//
//  Created by Victor Belov on 24/05/16.
//  Copyright © 2016 Victor Belov. All rights reserved.
//

#import "OpenHABNotificationsViewControllerTableViewController.h"
#import "OpenHABAppDataDelegate.h"
#import "OpenHABDataObject.h"
#import "NSMutableURLRequest+Auth.h"
#import "AFNetworking.h"
#import "AFRememberingSecurityPolicy.h"
#import "OpenHABNotification.h"
#import "NSDate+Helper.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import <SDWebImage/SDWebImageDownloader.h>
#import "UIViewController+MMDrawerController.h"
#import "MMDrawerBarButtonItem.h"
#import "OpenHABDrawerTableViewController.h"
#import "NotificationTableViewCell.h"


@interface OpenHABNotificationsViewControllerTableViewController ()

@end

@implementation OpenHABNotificationsViewControllerTableViewController

@synthesize openHABUsername, openHABRootUrl, openHABPassword, ignoreSSLCertificate, notifications;

- (void)viewDidLoad {
    [super viewDidLoad];
    self.notifications = [NSMutableArray array];
    self.tableView.tableFooterView = [[UIView alloc] init];
    self.refreshControl = [[UIRefreshControl alloc] init];
    self.refreshControl.backgroundColor = [UIColor groupTableViewBackgroundColor];
    //    self.refreshControl.tintColor = [UIColor whiteColor];
    [self.refreshControl addTarget:self action:@selector(handleRefresh:) forControlEvents:UIControlEventValueChanged];
    [self.tableView addSubview:self.refreshControl];
    [self.tableView sendSubviewToBack:self.refreshControl];
    MMDrawerBarButtonItem * rightDrawerButton = [[MMDrawerBarButtonItem alloc] initWithTarget:self action:@selector(rightDrawerButtonPress:)];
    [self.navigationItem setRightBarButtonItem:rightDrawerButton animated:YES];

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self loadSettings];
    [self loadNotifications];
}

- (void)loadNotifications {
    NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    NSString *notificationsUrlString = [NSString stringWithFormat:@"%@/api/v1/notifications?limit=20", [prefs valueForKey:@"remoteUrl"]];
    NSURL *notificationsUrl = [[NSURL alloc] initWithString:notificationsUrlString];
    NSMutableURLRequest *notificationsRequest = [NSMutableURLRequest requestWithURL:notificationsUrl];
    [notificationsRequest setAuthCredentials:self.openHABUsername :self.openHABPassword];
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:notificationsRequest];
    AFRememberingSecurityPolicy *policy = [AFRememberingSecurityPolicy policyWithPinningMode:AFSSLPinningModeNone];
    operation.securityPolicy = policy;
    if (self.ignoreSSLCertificate) {
        NSLog(@"Warning - ignoring invalid certificates");
        operation.securityPolicy.allowInvalidCertificates = YES;
    }
    operation.responseSerializer = [AFJSONResponseSerializer serializer];
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        NSData *response = (NSData*)responseObject;
        NSError *error;
        [self.notifications removeAllObjects];
        NSLog(@"Notifications response");
        // If we are talking to openHAB 1.X, talk XML
        if ([responseObject isKindOfClass:[NSArray class]]) {
            NSLog(@"Response is array");
            for (id notificationJson in responseObject) {
                OpenHABNotification *notification = [[OpenHABNotification alloc] initWithDictionary:notificationJson];
                [notifications addObject:notification];
            }
        } else {
            NSLog(@"Response is not array");
            return;
        }
        [self.refreshControl endRefreshing];
        [self.tableView reloadData];
        [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    } failure:^(AFHTTPRequestOperation *operation, NSError *error){
        [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
        NSLog(@"Error:------>%@", [error description]);
        NSLog(@"error code %ld",(long)[operation.response statusCode]);
        [self.refreshControl endRefreshing];
    }];
    [operation start];
}

- (void)handleRefresh:(UIRefreshControl *)refreshControl {
    NSLog(@"Refresh pulled");
    [self loadNotifications];
}

-(void)rightDrawerButtonPress:(id)sender{
    OpenHABDrawerTableViewController *drawer = (OpenHABDrawerTableViewController*)[self.mm_drawerController rightDrawerViewController];
    [self.mm_drawerController toggleDrawerSide:MMDrawerSideRight animated:YES completion:nil];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [notifications count];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"NotificationCell";
    NotificationTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    OpenHABNotification *notification = [notifications objectAtIndex:[indexPath row]];
    cell.textLabel.text = notification.message;
    // First convert date of notification from UTC from my.OH to local time for device
    NSTimeInterval timeZoneSeconds = [[NSTimeZone localTimeZone] secondsFromGMT];
    NSDate *createdInLocalTimezone = [notification.created dateByAddingTimeInterval:timeZoneSeconds];
    cell.detailTextLabel.text = [NSDate stringForDisplayFromDate:createdInLocalTimezone];
    
    NSString *iconUrlString = nil;
    if ([self appData].openHABVersion == 2) {
        iconUrlString = [NSString stringWithFormat:@"%@/icon/%@.png", [self appData].openHABRootUrl, notification.icon];
    } else {
        iconUrlString = [NSString stringWithFormat:@"%@/images/%@.png", [self appData].openHABRootUrl, notification.icon];
    }
    NSLog(@"%@", iconUrlString);
    [cell.imageView sd_setImageWithURL:[NSURL URLWithString:iconUrlString] placeholderImage:[UIImage imageNamed:@"icon-29x29.png"] options:0];
    if ([cell respondsToSelector:@selector(setPreservesSuperviewLayoutMargins:)]) {
        [cell setPreservesSuperviewLayoutMargins:NO];
    }
    // Explictly set your cell's layout margins
    if ([cell respondsToSelector:@selector(setLayoutMargins:)]) {
        [cell setLayoutMargins:UIEdgeInsetsZero];
    }
    cell.separatorInset = UIEdgeInsetsMake(0, 60, 0, 0);
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // open a alert with an OK and cancel button
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}

- (void)loadSettings
{
    NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
    self.openHABUsername = [prefs valueForKey:@"username"];
    self.openHABPassword = [prefs valueForKey:@"password"];
    self.ignoreSSLCertificate = [prefs boolForKey:@"ignoreSSL"];
//    self.defaultSitemap = [prefs valueForKey:@"defaultSitemap"];
//    self.idleOff = [prefs boolForKey:@"idleOff"];
    [[self appData] setOpenHABUsername:self.openHABUsername];
    [[self appData] setOpenHABPassword:self.openHABPassword];
}

- (OpenHABDataObject*)appData
{
    id<OpenHABAppDataDelegate> theDelegate = (id<OpenHABAppDataDelegate>) [UIApplication sharedApplication].delegate;
    return [theDelegate appData];
}

@end

File: openHAB/OpenHABNotificationsViewControllerTableViewController.swift
// Copyright (c) 2010-2019 Contributors to the openHAB project
//
// See the NOTICE file(s) distributed with this work for additional
// information.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0
//
// SPDX-License-Identifier: EPL-2.0

import SDWebImage
import UIKit

class OpenHABNotificationsViewController: UITableViewController {
    static let tableViewCellIdentifier = "NotificationCell"

    var notifications: NSMutableArray = []
    var openHABRootUrl = ""
    var openHABUsername = ""
    var openHABPassword = ""

    override func viewDidLoad() {
        super.viewDidLoad()
        notifications = []
        tableView.tableFooterView = UIView()
        refreshControl = UIRefreshControl()
        refreshControl?.backgroundColor = UIColor.groupTableViewBackground
        //    self.refreshControl.tintColor = [UIColor whiteColor];
        refreshControl?.addTarget(self, action: #selector(OpenHABNotificationsViewController.handleRefresh(_:)), for: .valueChanged)
        if let refreshControl = refreshControl {
            tableView.addSubview(refreshControl)
        }
        if let refreshControl = refreshControl {
            tableView.sendSubviewToBack(refreshControl)
        }
        let rightDrawerButton = MMDrawerBarButtonItem(target: self, action: #selector(OpenHABNotificationsViewController.rightDrawerButtonPress(_:)))
        navigationItem.setRightBarButton(rightDrawerButton, animated: true)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        loadSettings()
        loadNotifications()
    }

    func loadNotifications() {
        let prefs = UserDefaults.standard
        UIApplication.shared.isNetworkActivityIndicatorVisible = true
        var notificationsUrlString: String?
        if let value = prefs.value(forKey: "remoteUrl") {
            notificationsUrlString = "\(value)/api/v1/notifications?limit=20"
        }
        let notificationsUrl = URL(string: notificationsUrlString ?? "")
        var notificationsRequest: NSMutableURLRequest?
        if let notificationsUrl = notificationsUrl {
            notificationsRequest = NSMutableURLRequest(url: notificationsUrl)
        }
        var operation: OpenHABHTTPRequestOperation?
        if let notificationsRequest = notificationsRequest {
            operation = OpenHABHTTPRequestOperation(request: notificationsRequest as URLRequest)
        }
        operation?.responseSerializer = AFJSONResponseSerializer()
        operation?.setCompletionBlockWithSuccess({ _, responseObject in
            let response = responseObject as? Data
            self.notifications = []
            print("Notifications response")
            // If we are talking to openHAB 1.X, talk XML
            if response is [Any] {
                print("Response is array")
                for notificationJson: Any? in responseObject as! [Any?] {
                    let notification = OpenHABNotification(dictionary: notificationJson as! [AnyHashable: Any] as! [String: Any])
                    self.notifications.add(notification)
                }
            } else {
                print("Response is not array")
                return
            }
            self.refreshControl?.endRefreshing()
            self.tableView.reloadData()
            UIApplication.shared.isNetworkActivityIndicatorVisible = false
        }, failure: { operation, error in
            UIApplication.shared.isNetworkActivityIndicatorVisible = false
            print("Error:------>\(error.localizedDescription)")
            print(String(format: "error code %ld", Int(operation.response?.statusCode ?? 0)))
            self.refreshControl?.endRefreshing()
        })
        operation?.start()
    }

    @objc
    func handleRefresh(_ refreshControl: UIRefreshControl?) {
        print("Refresh pulled")
        loadNotifications()
    }

    @objc
    func rightDrawerButtonPress(_ sender: Any?) {
        mm_drawerController.toggle(MMDrawerSide.right, animated: true, completion: nil)
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return notifications.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: OpenHABNotificationsViewController.tableViewCellIdentifier) as? NotificationTableViewCell
        let notification = notifications[indexPath.row] as? OpenHABNotification
        cell?.textLabel?.text = notification?.message
        // First convert date of notification from UTC from my.OH to local time for device
        let timeZoneSeconds = TimeInterval(NSTimeZone.local.secondsFromGMT())
        let createdInLocalTimezone = notification?.created?.addingTimeInterval(timeZoneSeconds)
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.S'Z'"
        cell?.detailTextLabel?.text = dateFormatter.string(from: createdInLocalTimezone!)

        var iconUrlString: String?
        if appData()?.openHABVersion == 2 {
            if let app = appData()?.openHABRootUrl, let icon = notification?.icon {
                iconUrlString = "\(app)/icon/\(icon).png"
            }
        } else {
            if let app = appData()?.openHABRootUrl, let icon = notification?.icon {
                iconUrlString = "\(app)/images/\(icon).png"
            }
        }
        print("\(iconUrlString ?? "")")
        cell?.imageView?.sd_setImage(with: URL(string: iconUrlString ?? ""), placeholderImage: UIImage(named: "icon-29x29.png"), options: [])
        if cell?.responds(to: #selector(setter: NotificationTableViewCell.preservesSuperviewLayoutMargins)) ?? false {
            cell?.preservesSuperviewLayoutMargins = false
        }
        // Explictly set your cell's layout margins
        if cell?.responds(to: #selector(setter: NotificationTableViewCell.layoutMargins)) ?? false {
            cell?.layoutMargins = .zero
        }
        cell?.separatorInset = UIEdgeInsets(top: 0, left: 60, bottom: 0, right: 0)
        return cell!
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        // open a alert with an OK and cancel button
        tableView.deselectRow(at: indexPath, animated: false)
    }

    func loadSettings() {
        let prefs = UserDefaults.standard
        openHABUsername = prefs.value(forKey: "username") as? String ?? ""
        openHABPassword = prefs.value(forKey: "password") as? String ?? ""
        //    self.defaultSitemap = [prefs valueForKey:@"defaultSitemap"];
        //    self.idleOff = [prefs boolForKey:@"idleOff"];
        appData()?.openHABUsername = openHABUsername
        appData()?.openHABPassword = openHABPassword
    }

    func appData() -> OpenHABDataObject? {
        let theDelegate = UIApplication.shared.delegate as? OpenHABAppDataDelegate?
        return theDelegate??.appData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
